### DISCUSSION TITLE
Problem with class that inherit from Sprite

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1532] [General]
Hi,

Im with a problem that i don't understand what is wrong.

I have a class Weapon that inherit from Sprite:

WEAPON

```
#pragma once

#include "axmol.h"

namespace game::weapon
{

USING_NS_AX;

class Weapon : public Sprite
{
public:
    virtual Weapon *createWeapon() = 0;
    virtual void initializeData();
};

} // namespace game::weapon
```

BOMB

```
#pragma once

#include "weapons/Weapon.h"

namespace game::weapon
{

class Bomb : public Weapon
{
public:
    virtual Weapon *createWeapon() override;
    virtual void initializeData() override;
};

} // namespace game::weapon
```

BOMB IMPL

```
Weapon *Bomb::createWeapon()
{
    Weapon *weapon = static_cast<Weapon *>(Weapon::create("weapons/bomb/body.png"));
    weapon->initializeData(); // <--- CRASH HERE
    return weapon;
}

void Bomb::initializeData()
{
    setName("shoot");
    setScale(1.0f);
}
```

WEAPON IMPL

```
#include "weapons/Weapon.h"

namespace game::weapon
{

void Weapon::initializeData()
{
    //
}

} // namespace game::weapon
```

But when i try create a `bomb` and call their method `initializeData` it crash:

<img width="1316" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/ca97b374-ef67-4861-8615-5af165fe6669">

It is not null and when i comment the line `weapon->initializeData();` it works without crash.

What im doing wrong?

Thanks.

**Comment by aismann** [comment]
What about your WEAPON IMPL ?

**Reply by paulocoutinhox** [reply]
I update the thread with it, has nothing.

**Comment by rh101** [comment]
A `Sprite` is not a `Weapon`, yet your code is trying to force it to be. 

```cpp
Weapon *Bomb::createWeapon()
{
    Weapon *weapon = static_cast<Weapon *>(Weapon::create("weapons/bomb/body.png"));  <== incorrect logic
    weapon->initializeData(); // <--- CRASH HERE
    return weapon;
}
```

A `Weapon` is a `Sprite`, but a `Sprite` is not necessarily a `Weapon`.  You can cast from `Weapon` to `Sprite`, and vice-versa, if and only if that object is an instance of `Weapon`.  When downcasting, prefer `dynamic_cast` over `static_cast`. A dynamic cast would do the necessary checks to ensure that the object is in fact an instance of the type you are casting to, otherwise it will result in a `nullptr`, which you can check for.  With a `static_cast`, you wouldn't get the `nullptr` (as you noticed with the crash in your code).

It looks like you forgot to implement the static method `Weapon::create()`.  You can't use the `Sprite::create()`, since it's creating an object of type Sprite (the base class).

**Reply by rh101** [reply]
Also, having `createWeapon` as a virtual method makes little sense. It would mean you need to have an instance of that weapon in order to create another instance.  Unless, perhaps, your use-case is weapon spawning more copies of itself?

The correct logic is to have static `create()` convenience methods defined for each derived type.

**Comment by aismann** [comment]
> 
> I have a class Weapon that inherit from Sprite:
> 
Better:
Weapon using a Sprite object and isn't a Sprite itself
